### PR TITLE
feat: blstrs backend and optimized proving

### PR DESCRIPTION
**Pull request by dignifiedquire**
Uses https://github.com/dignifiedquire/blstrs to provide a `pairing` alternative backed by `blst`. 
Also integrates the optimizations done in https://github.com/filecoin-project/fil-blst/

### CODE REVIEW

File: src/groth16/tests/dummy_engine.rs [hunk 1]
```diff
@@ -281,7 +281,7 @@ impl Engine for DummyEngine {
         for &(a, b) in i {
             let mut tmp = *a;
             <Fr as Field>::mul_assign(&mut tmp, b);
-            <Fr as Field>::add_assign(&mut acc, &tmp);
+            <Fr as Field>::mul_assign(&mut acc, &tmp);
```
Comment (porcuquine) [inline src/groth16/tests/dummy_engine.rs:284]: What is this change about?

File: src/groth16/verifier.rs [hunk 2]
```diff
@@ -58,147 +47,221 @@ pub fn verify_proof<'a, E: Engine>(
...
+        rayon::scope(|s| {
+            // - Thread 1: Calculate ML alpha * beta
+            let ml_a_b = &mut ml_a_b;
+            s.spawn(move |_| {
+                *ml_a_b = E::miller_loop(&[(&proof.a.prepare(), &proof.b.prepare())]);
+            });
+
+            // - Thread 2: Calculate ML c * (-gamma)
+            let ml_all = &mut ml_all;
+            s.spawn(move |_| *ml_all = E::miller_loop(&[(&proof.c.prepare(), &pvk.neg_delta_g2)]));
```
Comment (porcuquine) [inline src/groth16/verifier.rs:68]: The comment says `-gamma`, but the code says `neg_delta_g2`.

Comment (dignifiedquire) [reply src/groth16/verifier.rs:68]: typo, fixed

File: src/groth16/tests/dummy_engine.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/groth16/tests/dummy_engine.rs]: reverted

**Review by arielgabizon — APPROVED** [review]
Went over changes with @dignifiedquire  and @porcuquine. Seem completely fine - a straightforward paralellization when possible of group operations.
There is a clash with the dummy engine implementation, but it has to do with the dummy engine using vector inner product as a miller loop, which is not really a miller loop - as it doesn't have the property in particular that e(ML(A,B)*ML(C,D)) = e(A,B)*e(C,D)  where e is the pairing and ML is the miller loop.

**Review by porcuquine — APPROVED** [review]
Based on discussion with @arielgabizon and on @dignifiedquire's reversion of the change to the `DummyEngine` implementation, I approve. Let's create an issue to make sure we follow up with a clear fix to that test eventually.

**Comment by cryptonemo** [comment]
Is this CI failure an expected CI timeout where we could extend the test run time or a real bug holding up this work?

**Comment by porcuquine** [comment]
No, we need to sort out what is happening. I'm looking into it but don't have an answer or fix yet.

**Pull request closed without merging**